tests/trainer/test_data_loading.py

```diff
@@ -254,3 +254,62 @@ class CustomSampler(Sampler):
     dataloader = CustomDataLoader(dataset, sampler=CustomSampler(dataset))
     with pytest.raises(MisconfigurationException, match="will be replaced  by `DistributedSampler`"):
         trainer.auto_add_sampler(dataloader, shuffle=True)
+
+
+def test_loader_detaching():
+    """Checks that the loader has been resetted after the entrypoint"""
+
+    class LoaderTestModel(BoringModel):
+        def training_step(self, batch, batch_idx):
+            assert len(model.train_dataloader()) == 10
+            return super().training_step(batch, batch_idx)
+
+        def validation_step(self, batch, batch_idx):
+            assert len(model.val_dataloader()) == 10
+            return super().validation_step(batch, batch_idx)
+
+        def test_step(self, batch, batch_idx):
+            assert len(model.test_dataloader()) == 10
+            return super().test_step(batch, batch_idx)
+
+        def predict_step(self, batch, batch_idx, dataloader_idx=None):
+            assert len(model.predict_dataloader()) == 10
+            return super().predict_step(batch, batch_idx, dataloader_idx=dataloader_idx)
+
+    loader = DataLoader(RandomDataset(32, 10), batch_size=1)
+
+    model = LoaderTestModel()
+
+    assert len(model.train_dataloader()) == 64
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
+
+    trainer = Trainer(fast_dev_run=1)
+    trainer.fit(model, loader, loader)
+
+    assert len(model.train_dataloader()) == 64
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
+
+    trainer.validate(model, loader)
+
+    assert len(model.train_dataloader()) == 64
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
+
+    trainer.predict(model, loader)
+
+    assert len(model.train_dataloader()) == 64
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
+
+    trainer.test(model, loader)
+
+    assert len(model.train_dataloader()) == 64
+    assert len(model.val_dataloader()) == 64
+    assert len(model.predict_dataloader()) == 64
+    assert len(model.test_dataloader()) == 64
```

Comment:
Would this be enough to test?

(i have omitted `validate` because it gets set with `fit`)

```suggestion
    trainer.fit(model, loader, loader)
    trainer.predict(model, loader)
    trainer.test(model, loader)

    assert len(model.train_dataloader()) == 64
    assert len(model.val_dataloader()) == 64
    assert len(model.predict_dataloader()) == 64
    assert len(model.test_dataloader()) == 64
```

Reply:
This would still catch it, but I explicitly wanted to have tests for all the entrypoints that each of them resets all loaders. We can keep it that way though.